app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt

```diff
@@ -1333,6 +1337,80 @@ class AuthRepositoryImpl(
     }
 
+    override fun getNewDeviceNoticeState(): NewDeviceNoticeState? {
+        return activeUserId?.let { userId ->
+            authDiskSource.getNewDeviceNoticeState(userId = userId)
+        }
+    }
+
+    override fun setNewDeviceNoticeState(newState: NewDeviceNoticeState?) {
+        activeUserId?.let { userId ->
+            authDiskSource.storeNewDeviceNoticeState(userId = userId, newState = newState)
+        }
+    }
+
+    override fun checkUserNeedsNewDeviceTwoFactorNotice(): Boolean {
+        return activeUserId?.let { userId ->
+            val temporaryFlag = featureFlagManager.getFeatureFlag(FlagKey.NewDeviceTemporaryDismiss)
+            val permanentFlag = featureFlagManager.getFeatureFlag(FlagKey.NewDevicePermanentDismiss)
+
+            // check if feature flags are disabled
+            if (!temporaryFlag && !permanentFlag) {
+                return false
+            }
+
+            if (!newDeviceNoticePreConditionsValid()) {
+                return false
+            }
+
+            val newDeviceNoticeState = authDiskSource.getNewDeviceNoticeState(userId = userId)
+            return when (newDeviceNoticeState.displayStatus) {
+                // if the user has already attested email access but permanent flag is enabled,
+                // the notice needs to appear again
+                NewDeviceNoticeDisplayStatus.CAN_ACCESS_EMAIL -> permanentFlag
+                // if the user has already seen but 7 days have already passed,
+                // the notice needs to appear again
+                NewDeviceNoticeDisplayStatus.HAS_SEEN ->
+                    newDeviceNoticeState.delayDateIsSevenDaysOld
+                NewDeviceNoticeDisplayStatus.HAS_NOT_SEEN -> true
+                // the user never needs to see the notice again
+                NewDeviceNoticeDisplayStatus.CAN_ACCESS_EMAIL_PERMANENT -> false
+            }
+        }
+            ?: false
+    }
+
+    private fun newDeviceNoticePreConditionsValid(): Boolean {
+        val hasSSOPolicy =
+            policyManager.getActivePolicies(type = PolicyTypeJson.REQUIRE_SSO)
+                .any { p -> p.isEnabled }
+        val isCloudUser = environmentRepository.environment.type != Environment.Type.SELF_HOSTED
+        val userProfile = authDiskSource.userState?.activeAccount?.profile
+        val hasTwoFactorEnabled = userProfile
+            ?.isTwoFactorEnabled
+            ?: false
+        val isProfileAtLeastWeekOld = userProfile
+            ?.let {
+                it.creationDate
+                    ?.plusWeeks(1)
+                    ?.isBefore(
+                        ZonedDateTime.now(),
+                    )
+            }
+            ?: false
+
+        // broken down for easier readability
+        if (!isCloudUser || !isProfileAtLeastWeekOld) {
+            return false
+        }
+
+        if (hasSSOPolicy || hasTwoFactorEnabled) {
+            return false
+        }
+
+        return true
+    }
```

Comment:
🎨  Up to you but IMO you could short-circuit it to return in several parts this so you don't need to waste resources fetching other stuff. Like there's no need to fetch the policies if the user is not cloud.
```suggestion
    private fun newDeviceNoticePreConditionsValid(): Boolean {
        if (environmentRepository.environment.type == Environment.Type.SELF_HOSTED) {
            return false
        }

        val userProfile = authDiskSource.userState?.activeAccount?.profile
        val isProfileAtLeastWeekOld = userProfile
            ?.let {
                it.creationDate
                    ?.plusWeeks(1)
                    ?.isBefore(
                        ZonedDateTime.now(),
                    )
            }
            ?: false
        if (!isProfileAtLeastWeekOld) {
            return false
        }

        val hasTwoFactorEnabled = userProfile
            ?.isTwoFactorEnabled
            ?: false
        if (hasTwoFactorEnabled) {
            return false
        }

        val hasSSOPolicy =
            policyManager.getActivePolicies(type = PolicyTypeJson.REQUIRE_SSO)
                .any { p -> p.isEnabled }

        return !hasSSOPolicy
    }
```